api/src/auth/api_user_key_auth.py

```diff
@@ -0,0 +1,77 @@
+import logging
+
+from apiflask import HTTPTokenAuth
+from flask import current_app
+from sqlalchemy import select
+from sqlalchemy.orm import selectinload
+
+import src.util.datetime_util as datetime_util
+from src.adapters import db
+from src.adapters.db import flask_db
+from src.api.route_utils import raise_flask_error
+from src.db.models.user_models import UserApiKey
+from src.logging.flask_logger import add_extra_data_to_current_request_logs
+
+logger = logging.getLogger(__name__)
+
+# Initialize the authorization context for API Gateway key authentication
+# This uses the X-API-Key header which is the standard header that AWS API Gateway
+# forwards when api_key_required is set to true
+api_user_key_auth = HTTPTokenAuth(
+    "ApiKey", header="X-API-Key", security_scheme_name="ApiUserKeyAuth"
+)
+
+
+class ApiKeyValidationError(Exception):
+    """Exception raised when API key validation fails"""
+
+    def __init__(self, message: str):
+        self.message = message
+        super().__init__(message)
+
+
+@api_user_key_auth.verify_token
+def verify_api_key(token: str) -> UserApiKey:
+    logger.info("Authenticating API Gateway key")
+
+    # Get database session
+    with flask_db.get_db(current_app).get_session() as db_session:
+        try:
+            api_key = validate_api_key_in_db(token, db_session)
+            api_key.last_used = datetime_util.utcnow()
+            db_session.add(api_key)
+            db_session.commit()
+
+            add_extra_data_to_current_request_logs(
+                {
+                    "auth.user_id": str(api_key.user_id),
+                    "auth.api_key_id": str(api_key.api_key_id),
+                }
+            )
+
+            logger.info("API Gateway key authentication successful")
+
+            return api_key
+
+        except ApiKeyValidationError as e:
+            logger.info("API Gateway key authentication failed", extra={"auth.issue": e.message})
+            raise_flask_error(401, e.message)
+
+
+def validate_api_key_in_db(api_key: str, db_session: db.Session) -> UserApiKey:
+    user_api_key: UserApiKey | None = db_session.execute(
+        select(UserApiKey)
+        .where(UserApiKey.key_id == api_key)
+        .options(selectinload(UserApiKey.user))
+    ).scalar_one_or_none()
+
+    if user_api_key is None:
+        raise ApiKeyValidationError("Invalid API key")
+
+    if not user_api_key.is_active:
+        raise ApiKeyValidationError("API key is inactive")
+
+    if user_api_key.user is None:
+        raise ApiKeyValidationError("API key is not associated with a valid user")
```

Comment:
User ID (and thus the user relationship) are non-nullable and the foreign key constraint enforces that it'll point to something, can skip this check.

```suggestion
```